eole/inputters/text_utils.py

```diff
@@ -87,19 +89,33 @@ def numericalize(vocabs, example, model_type=ModelType.ENCODER_DECODER):
                 )
 
     elif model_type == ModelType.DECODER:
-        if numeric["src"]["src_ids"] == []:
+        if numeric["src"]["src_ids"] == []:  # tokenize_id not used
             src_text = example["src"]["src"].split(" ")
             if decoder_start_token != "":
                 src_text = [decoder_start_token] + src_text
             numeric["src"]["src_ids"] = vocabs["src"](src_text)
-        if example["tgt"] is not None:
+        numeric["src"]["prefix_len"] = len(numeric["src"]["src_ids"])
+        if example["tgt"] is None:  # no path_tgt given need to use src
+            if task == CorpusTask.TRAIN:
+                example["tgt"] = {}
+                if maybe_tgt_ids != []:
+                    numeric["tgt"]["tgt_ids"] = maybe_tgt_ids
+                else:
+                    tgt_text = example["src"]["src"].split(" ")
+                    numeric["tgt"]["tgt_ids"] = vocabs["tgt"](tgt_text + [vocabs["specials"].get("eos_token", "")])
+                if decoder_start_token == "":
+                    numeric["tgt"]["tgt_ids"] = numeric["tgt"]["tgt_ids"][1:]
+        else:  # path_tgt given = answer to prompt from path_src
             if maybe_tgt_ids != []:
-                numeric["tgt"]["tgt_ids"] = maybe_tgt_ids
+                numeric["tgt"]["tgt_ids"] = numeric["src"]["src_ids"] + maybe_tgt_ids
+                numeric["src"]["src_ids"] = numeric["src"]["src_ids"] + maybe_tgt_ids[:-1]
             else:
                 tgt_text = example["tgt"]["tgt"].split(" ")
                 numeric["tgt"]["tgt_ids"] = vocabs["tgt"](tgt_text + [vocabs["specials"].get("eos_token", "")])
-            if decoder_start_token == "":
-                numeric["tgt"]["tgt_ids"] = numeric["tgt"]["tgt_ids"][1:]
+                numeric["src"]["src_ids"] = numeric["src"]["src_ids"] + numeric["tgt"]["tgt_ids"]
+                numeric["tgt"]["tgt_ids"] = numeric["src"]["src_ids"]
+                numeric["src"]["src_ids"] = numeric["src"]["src_ids"][:-1]
+            numeric["tgt"]["tgt_ids"] = numeric["tgt"]["tgt_ids"][1:]
```

Comment:
This is less and less readable. Might be good to add a few comments to clarify the various cases, offsets, etc. while things are fresh.